### PR TITLE
Validate new m.room.power_levels events

### PR DESCRIPTION
**Pull request by aaronraimist**
Fixes https://github.com/matrix-org/synapse/issues/1237

I'm planning on writing a test but I have not written one yet. Is Complement the right place to create that now?

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [ ] Pull request includes a [changelog file](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
* [x] Pull request includes a [sign off](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#sign-off)
* [x] Code style is correct (run the [linters](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#code-style))

### CODE REVIEW

**Comment by aaronraimist** [comment]
I'm not sure what is up with the complement test failure. Something to do with spaces?

**Comment by erikjohnston** [comment]
Hey, thanks for this! Looks good, can I ask that you add a couple of basic tests for this? Hopefully should be easy to do

File: synapse/events/validator.py [hunk 1]
```diff
@@ -185,3 +237,8 @@ def _ensure_strings(self, d, keys):
     def _ensure_state_event(self, event):
         if not event.is_state():
             raise SynapseError(400, "'%s' must be state events" % (event.type,))
+
+    def _is_object(self, checker, instance):
+        if isinstance(instance, (dict, frozendict.frozendict)):
```
Comment (richvdh) [inline synapse/events/validator.py:242]: we normally write this:

```suggestion
        if isinstance(instance, collections.abc.Mapping):
```

File: synapse/python_dependencies.py [hunk 2]
```diff
@@ -48,7 +48,7 @@
 # [1] https://pip.pypa.io/en/stable/reference/pip_install/#requirement-specifiers.
 
 REQUIREMENTS = [
-    "jsonschema>=2.5.1",
+    "jsonschema>=3.0.0",
```
Comment (richvdh) [inline synapse/python_dependencies.py:51]: could you add a comment here to record why we need 3.0 (in a similar way to the other comments below)

File: synapse/events/validator.py [hunk 3]
```diff
@@ -87,6 +90,55 @@ def validate_new(self, event: EventBase, config: HomeServerConfig):
...
+                    "redact": {"$ref": "#/definitions/int"},
+                    "state_default": {"$ref": "#/definitions/int"},
+                    "users": {"$ref": "#/definitions/objectOfInts"},
+                    "users_default": {"$ref": "#/definitions/int"},
+                },
+                "definitions": {
+                    "int": {
+                        "type": "integer",
+                        "minimum": -9007199254740991,
+                        "maximum": 9007199254740991,
```
Comment (richvdh) [inline synapse/events/validator.py:112]: could I ask you to factor these out to constants somewhere (eg, `CANONICALJSON_{MIN,MAX}_INT`)? There is another reference in `synapse.events.utils` which could do with de-duplicating.

**Review by richvdh — CHANGES_REQUESTED** [review]
this looks great, thanks! A few minor requests from me.

For this sort of thing, some python unit tests are generally an easier way of exercising various edge cases, so it would be great to see some. I won't insist though, because the complement test is certainly useful.

**Comment by anoadragon453** [comment]
Hey @aaronraimist, are you able to continue working on this? :slightly_smiling_face:

**Comment by aaronraimist** [comment]
@anoadragon453 I've been a bit busy but I'll try to get back to this in the next few days. (if anyone wants to just swoop in and fix it though by all means go ahead)

**Comment by aaronraimist** [comment]
Sorry for the delay. I think this is ready for another review.
